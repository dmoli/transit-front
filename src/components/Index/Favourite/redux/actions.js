import { set as setRoute } from '../../Route/redux/actions';

/**
 * Action types
 */
export const actionTypes = {
  /** Set */
  SET_FAVOURITE: 'SET_FAVOURITE',
};

/**
 * Modify entities
 *
 * @param {array} entities entities
 * @return action to dispatch
 */
export const set = entities => ({
  type: actionTypes.SET_FAVOURITE,
  payload: entities,
});

/**
 * Find by id
 *
 * @param {int} routeId route id
 * @param {array} routes entities
 * @return entity found
 */
const findById = (routeId, routes) => {
  const setedRoutes = JSON.parse(JSON.stringify(routes));
  let entitySeted;
  for (const entity of setedRoutes) {
    if (routeId === entity.route_id) {
      entitySeted = entity;
      break;
    }
  }
  return entitySeted;
};

/**
 * Set object fields of the array by actions
 *
 * @param {int} id id of the entity
 * @param {array} entities entities
 * @param {string} action action to do
 * @return {array} setedEntities entities seted
 */
const setFields = (id, entities, action) => {
  const setedEntities = JSON.parse(JSON.stringify(entities));
  let i = 0;
  for (const entity of setedEntities) {
    if (id === entity.route_id) {
      const entitySeted = entity;
      if (action === 'fav') entitySeted.favorited = true;
      if (action === 'unfav') entitySeted.favorited = false;
      setedEntities[i] = entitySeted;
      break;
    }
    i += 1;
  }
  return setedEntities;
};

/**
 * Add favorite - Mock only
 *
 * @param {int} routeId route id
 * @param {object} dispatch dispatch of actions
 * @return action to dispatch
 */
export const add = routeId => (
  async (dispatch, getState) => {
    const state = getState();
    const route = findById(routeId, state.routes.entities);
    route.favorited = true;
    const favourites = state.favourites.entities;
    favourites.push(route);
    // add this route to state
    dispatch(set(favourites));

    // refresh routes
    const routesSeted = setFields(routeId, state.routes.entities, 'fav');
    dispatch(setRoute(routesSeted));
  }
);

/**
 * Remove favorite - Mock only
 *
 * @param {int} routeId route id
 * @param {object} dispatch dispatch of actions
 * @return action to dispatch
 */
export const remove = routeId => (
  async (dispatch, getState) => {
    const state = getState();
    let i = 0;
    const setedFavourites = JSON.parse(JSON.stringify(state.favourites.entities));
    // find route in favourite and remove it
    for (const entity of setedFavourites) {
      if (routeId === entity.route_id) {
        // delete setedFavourites[i];
        setedFavourites.splice(i, 1);
        break;
      }
      i += 1;
    }
    console.log('setedFavourites', setedFavourites);
    // refresh favourites
    dispatch(set(setedFavourites));

    // refresh routes
    const routesSeted = setFields(routeId, state.routes.entities, 'unfav');
    dispatch(setRoute(routesSeted));
  }
);
