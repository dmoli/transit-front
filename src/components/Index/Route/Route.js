/* eslint no-return-assign: [0] */
import React, { Component } from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import {
  FaStar,
} from 'react-icons/lib/fa';
import {
  MdFavoriteOutline,
} from 'react-icons/lib/md';

/**
 * Componente ruta
 */
class Route extends Component {
  /**
   * Handle favourite actions
   *
   * @param {int} routeId route id
   * @param {string} action action name, favourite or unfavourite
   * @return array of routes seted
   */
  handleFavourites(routeId, action) {
    const { onClickToggleFavorite } = this.props;
    onClickToggleFavorite(routeId, action);
  }

  /**
   * Handle current behavior
   *
   * @param {int} routeId route id
   * @param {string} routeName route name
   * @param {bol} currentState current boolean field
   */
  handleCurrent(routeId, routeName, currentState) {
    if (currentState === true) return;
    const { onClickCurrent } = this.props;
    onClickCurrent(routeId, routeName, 'current');
  }

  render() {
    const { item } = this.props;

    return (
      <ContainerRoute
        key={item.route_id}
        background={item.route_color}
        color={item.route_text_color}
        current={item.current}
        onClick={() => this.handleCurrent(item.route_id, item.route_short_name, item.current)}
      >
        <ContainerNames>
          <Number>{item.route_short_name}</Number>
          <Name>{item.route_long_name}</Name>
        </ContainerNames>
        <Favourite
          onClick={() => this.handleFavourites(item.route_id, (item.favorited ? 'unfav' : 'fav'))}
        >
          {
            item.favorited
          }
          <MdFavoriteOutline
            size={30}
            color={'#d40066'}
            className={item.favorited ? 'pointer-hover red' : 'pointer-hover white'}
          />
        </Favourite>
      </ContainerRoute>
    );
  }
}

Route.propTypes = {
  /* item a mostrar */
  item: PropTypes.object.isRequired,
  /** function - active/desactive a favourite */
  onClickToggleFavorite: PropTypes.func.isRequired,
  /** function - active/desactive a current */
  onClickCurrent: PropTypes.func.isRequired,
};

const ContainerRoute = styled.section`
  display: flex;
  flex-direction: row;
  align-items: center;
  background-attachment: fixed;
  background-size: cover;
  background-position: center center;
  background-color: #${props => props.background || 'fff'};
  color: #${props => props.color || '000'};
  padding: 30px;
  margin-bottom: 15px;
  width: 60%;
  border-radius: 18px;
  box-shadow: 0px 0px 28px -6px #000;
  ${props => props.current === true && `
    background: #00f;
    color: #fff;
  `}
  /* &:hover {
    box-shadow: 0px 0px 11px 4px rgba(167, 166, 166, 0.42);
    cursor: pointer;
  } */
`;

const ContainerNames = styled.article`
  display: flex;
  flex-direction: column;
  width: 90%;
`;

const Number = styled.h1`
  font-size: 4em;
`;

const Name = styled.article``;

const Favourite = styled.div`
  border-radius: 50%;
  padding: 10px;
  /* box-shadow: 0px 0px 24px -6px #000;
  :hover {
    box-shadow: 0px 0px 36px -7px #000;
  } */
`;

export default Route;
