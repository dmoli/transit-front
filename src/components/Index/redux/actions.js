import * as api from '../../../api/routes';

/**
 * Action types
 */
export const actionTypes = {
  /** Set .entities */
  SET_ROUTES: 'SET_ROUTES',
  /** Set .entities && .page */
  SET_NEXT_PAGE_ROUTES: 'SET_NEXT_PAGE_ROUTES',
};

/**
 * Modify entities
 *
 * @param {array} entities entities
 * @return action to dispatch
 */
export const set = entities => ({
  type: actionTypes.SET_ROUTES,
  payload: entities,
});

/**
 * Modify entities and pagination +1
 *
 * @param {array} entities results
 * @return action to dispatch
 */
export const setNextPage = entities => ({
  type: actionTypes.SET_NEXT_PAGE_ROUTES,
  payload: entities,
});

/**
 * Get entities - API
 *
 * @param {object} dispatch dispatch of actions
 * @return action to dispatch
 */
export const get = () => (
  async (dispatch) => {
    try {
      // get response - API
      const response = await api.get();
      if (response.status !== 200) {
        throw new Error('error');
      }

      // response to json
      const data = await response.json();
      // dispatch action
      dispatch(set(data));
      // if not exist
      if (data.length === 0) {
        throw new Error('empty');
      }

      return data;
    } catch (e) {
      // error
      throw new Error(e.message);
    }
  }
);
