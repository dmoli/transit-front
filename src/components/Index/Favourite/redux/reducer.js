import { combineReducers } from 'redux';

import initialState from '../../../../redux/state';
import { actionTypes } from './actions';

/**
 * Entities reducer
 *
 * @param {array} state list of entities
 * @param action action to do
 */
const entitiesReducer = (state = initialState.favourites.entities, action = {}) => {
  switch (action.type) {
    case actionTypes.SET_FAVOURITE:
      return action.payload;
    default:
      return state;
  }
};


const reducer = combineReducers({
  entities: entitiesReducer,
});

export default reducer;
