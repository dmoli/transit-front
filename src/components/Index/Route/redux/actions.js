/* eslint no-restricted-syntax: [0] */
import * as api from '../../../../api/routes';
import { set as setFavourite } from '../../Favourite/redux/actions';

/**
 * Action types
 */
export const actionTypes = {
  /** Set .entities */
  SET_ROUTES: 'SET_ROUTES',
  /** Set .entities && .page */
  SET_NEXT_PAGE_ROUTES: 'SET_NEXT_PAGE_ROUTES',
  /** Set .current */
  SET_CURRENT: 'SET_CURRENT',
  /** Set .shapes */
  SET_SHAPES: 'SET_SHAPES',
};

/**
 * Modify entities
 *
 * @param {array} entities entities
 * @param {int} page optional number page
 * @return action to dispatch
 */
export const set = (entities, page = 1) => ({
  type: actionTypes.SET_ROUTES,
  payload: {
    entities,
    page,
  },
});

/**
 * Modify entities and pagination +1
 *
 * @param {array} entities results
 * @return action to dispatch
 */
export const setNextPage = entities => ({
  type: actionTypes.SET_NEXT_PAGE_ROUTES,
  payload: entities,
});

/**
 * Modify current
 *
 * @param {string} currentName currentName
 * @return action to dispatch
 */
export const setCurrent = currentName => ({
  type: actionTypes.SET_CURRENT,
  payload: currentName,
});

/**
 * Modify shapes
 *
 * @param {array} shapes shapes
 * @return action to dispatch
 */
export const setShapes = shapes => ({
  type: actionTypes.SET_SHAPES,
  payload: shapes,
});

/**
 * Verify if a route exist in favourites
 *
 * @param {array} routes entities
 * @param {array} favourites entities
 * @return {array} entities seted
 */
const setFavoriteState = (routes, favourites) => {
  const setedRoutes = JSON.parse(JSON.stringify(routes));
  const setedFavourites = JSON.parse(JSON.stringify(favourites));
  let i = 0;
  for (const routeItem of setedRoutes) {
    for (const favouriteItem of setedFavourites) {
      if (routeItem.id === favouriteItem.id) {
        const routeSetedItem = routeItem;
        routeSetedItem.favorited = true;
        setedRoutes[i] = routeSetedItem;
      }
    }
    i += 1;
  }
  return setedRoutes;
};

/**
 * Set object fields of the array by actions
 *
 * @param {int} id id of the entity
 * @param {array} entities entities
 * @param {string} action action to do
 * @return {array} setedEntities entities seted
 */
const setFields = (id, entities, action) => {
  const setedEntities = JSON.parse(JSON.stringify(entities));
  let i = 0;
  if (action === 'current') {
    for (const entity of setedEntities) {
      const entitySeted = entity;
      if (id === entity.route_id) {
        entitySeted.current = true;
      } else {
        entitySeted.current = false;
      }
      setedEntities[i] = entitySeted;
      i += 1;
    }
  }

  i = 0;
  if (action === 'init_shape') {
    for (const entity of setedEntities) {
      const entitySeted = entity;
      entitySeted.lat = parseFloat(entitySeted.shape_pt_lat);
      entitySeted.lng = parseFloat(entitySeted.shape_pt_lon);
      setedEntities[i] = entitySeted;
      i += 1;
    }
  }
  return setedEntities;
};

/**
 * Refresh current route
 *
 * @param {int} routeId route id
 * @param {string} routeName route name
 * @param {object} dispatch dispatch of actions
 * @return action to dispatch
 */
export const refreshCurrent = (routeId, routeName) => (
  async (dispatch, getState) => {
    const state = getState();
    // refresh routes
    const routesSeted = setFields(routeId, state.routes.entities, 'current');
    dispatch(set(routesSeted, state.routes.page));

    // refresh favourites
    const favouritesSeted = setFields(routeId, state.favourites.entities, 'current');
    dispatch(setFavourite(favouritesSeted));

    // refresh current field
    dispatch(setCurrent(routeName));
  }
);

/**
 * Get entities - API
 *
 * @param {string} text text to search
 * @param {object} dispatch dispatch of actions
 * @return action to dispatch
 */
export const get = (text = '', page = null) => (
  async (dispatch, getState) => {
    try {
      const state = getState();
      const pageToSearch = page === null ? state.routes.page : page;
      // get response - API
      const response = await api.get(pageToSearch, text);
      if (response.status !== 200) {
        throw new Error('error');
      }

      // response to json
      const data = await response.json();
      // verify if a route exist in favourites
      const dataSeted = setFavoriteState(data, state.favourites.entities);
      // dispatch action
      dispatch(set(dataSeted));
      // if not exist
      if (dataSeted.length === 0) {
        throw new Error('empty');
      }

      return dataSeted;
    } catch (e) {
      // error
      throw new Error(e.message);
    }
  }
);

/**
 * Get next page of entities - API
 *
 * @param {string} text text to search
 * @param {object} dispatch dispatch of actions
 * @return action to dispatch
 */
export const getNextPage = (text = '') => (
  async (dispatch, getState) => {
    try {
      const state = getState();
      // get response - API
      const response = await api.get(state.routes.page + 1, text);
      if (response.status !== 200) {
        throw new Error('error');
      }

      // response to json
      const data = await response.json();
      // if not exist
      if (data.length === 0) {
        throw new Error('empty');
      }

      // dispatch action
      dispatch(setNextPage(data));

      return data;
    } catch (e) {
      // error
      throw new Error(e.message);
    }
  }
);

/**
 * Get shapes - API
 *
 * @param {object} dispatch dispatch of actions
 * @return action to dispatch
 */
export const getShapes = routeId => (
  async (dispatch) => {
    try {
      // get response - API
      const response = await api.getShapes(routeId);
      if (response.status !== 200) {
        throw new Error('error');
      }

      // response to json
      const data = await response.json();
      // if not exist
      if (data.length === 0) {
        throw new Error('empty');
      }

      // set init
      const dataSeted = setFields(null, data, 'init_shape');
      // dispatch action
      dispatch(setShapes(dataSeted));

      return data;
    } catch (e) {
      // error
      throw new Error(e.message);
    }
  }
);
