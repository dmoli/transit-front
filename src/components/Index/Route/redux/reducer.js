import { combineReducers } from 'redux';

import initialState from '../../../../redux/state';
import { actionTypes } from './actions';

/**
 * Entities reducer
 *
 * @param {array} state list of entities
 * @param action action to do
 */
const entitiesReducer = (state = initialState.routes.entities, action = {}) => {
  switch (action.type) {
    case actionTypes.SET_ROUTES:
      return action.payload.entities;
    case actionTypes.SET_NEXT_PAGE_ROUTES:
      return state.concat(action.payload);
    default:
      return state;
  }
};

/**
 * Number page reducer
 *
 * @param {array} state number page
 * @param action plus +1 in the number page
 */
const pageReducer = (state = initialState.routes.page, action = {}) => {
  switch (action.type) {
    case actionTypes.SET_ROUTES:
      return action.payload.page;
    case actionTypes.SET_NEXT_PAGE_ROUTES:
      return state + 1;
    default:
      return state;
  }
};

/**
 * Current name reducer
 *
 * @param {array} state current name
 * @param action change name
 */
const currentReducer = (state = initialState.routes.current, action = {}) => {
  switch (action.type) {
    case actionTypes.SET_CURRENT:
      return action.payload;
    default:
      return state;
  }
};

/**
 * Shapes reducer
 *
 * @param {array} state list of shapes
 * @param action action to do
 */
const shapesReducer = (state = initialState.routes.shapes, action = {}) => {
  switch (action.type) {
    case actionTypes.SET_SHAPES:
      return action.payload;
    default:
      return state;
  }
};


const reducer = combineReducers({
  entities: entitiesReducer,
  page: pageReducer,
  current: currentReducer,
  shapes: shapesReducer,
});

export default reducer;
