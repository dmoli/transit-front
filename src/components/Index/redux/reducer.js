import { combineReducers } from 'redux';

import initialState from '../../../redux/state';
import { actionTypes } from './actions';

/**
 * Entities reducer
 *
 * @param {array} state list of entities
 * @param action action to do
 */
const entitiesReducer = (state = initialState.routes.entities, action = {}) => {
  switch (action.type) {
    case actionTypes.SET_ROUTES:
      return action.payload;
    case actionTypes.SET_NEXT_PAGE_ROUTES:
      return state.concat(action.payload);
    default:
      return state;
  }
};

/**
 * Number page reducer
 *
 * @param {array} state number page
 * @param action plus +1 in the number page
 */
const pageReducer = (state = initialState.routes.page, action = {}) => {
  switch (action.type) {
    case actionTypes.SET_ROUTES:
      return 1;
    case actionTypes.SET_NEXT_PAGE_ROUTES:
      return state + 1;
    default:
      return state;
  }
};


const reducer = combineReducers({
  entities: entitiesReducer,
  page: pageReducer,
});

export default reducer;
