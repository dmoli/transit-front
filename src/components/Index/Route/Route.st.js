/* eslint import/no-extraneous-dependencies: [0] */
import React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';
import { action } from '@storybook/addon-actions';
import { Provider } from 'react-redux';

import Route from './Route';
import { initStore } from '../../../redux/store';
import * as routeFixure from '../../../fixtures/route';

const routes = routeFixure.get();
const route = routes[0];
const store = initStore();

storiesOf('Index/Route', module)
  .addDecorator(story => <Provider store={store}>{story()}</Provider>)
  .add('Route',
    withInfo('Route')(() =>
      <Route
        item={route}
        onClickToggleFavorite={action('onClickToggleFavorite')}
        onClickCurrent={action('onClickCurrent')}
      />,
    ),
  );
