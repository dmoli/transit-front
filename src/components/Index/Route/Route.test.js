/* eslint-env jest */
import React from 'react';
import { Provider } from 'react-redux';

import Route from './Route';
import { mountWithIntl } from '../../../scripts/intl-enzyme-test-helper';
import { initStore } from '../../../redux/store';
import * as routeFixure from '../../../fixtures/route';

const routes = routeFixure.get();
const route = routes[0];
const store = initStore();

describe('<Route />', () => {
  it('should rendered correctly', () => {
    const component = mountWithIntl(
      <Provider store={store}>
        <Route
          item={route}
          onClickToggleFavorite={() => {}}
          onClickCurrent={() => {}}
        />
      </Provider>,
    );
    expect(component.html()).toMatchSnapshot();
  });
});
