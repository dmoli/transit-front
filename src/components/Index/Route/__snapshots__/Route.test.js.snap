// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Route /> should rendered correctly 1`] = `"<section class=\\"white-bg Route__ContainerRoute-sc-1apnz1v-0 eWrbSm\\" color=\\"000000\\"><article class=\\"Route__ContainerNames-sc-1apnz1v-1 bbNrow\\"><h1 class=\\"Route__Number-sc-1apnz1v-2 drdLtS\\" color=\\"00D5FF\\">101</h1><article class=\\"Route__Name-sc-1apnz1v-3 hCBzRs\\">Recoleta - Cerrillos</article></article><div class=\\"Route__Favourite-sc-1apnz1v-4 hqcHs\\"><svg fill=\\"currentColor\\" preserveAspectRatio=\\"xMidYMid meet\\" height=\\"30\\" width=\\"30\\" viewBox=\\"0 0 40 40\\" class=\\"pointer-hover\\" style=\\"vertical-align: middle; color: rgb(233, 0, 75);\\"><g><path d=\\"m20.2 30.9q3.7-3.3 5.5-5t3.9-4.1 2.9-4.1 0.9-3.6q0-2.5-1.7-4.1t-4.2-1.6q-2 0-3.6 1.1t-2.3 2.8h-3.2q-0.6-1.7-2.3-2.8t-3.6-1.1q-2.5 0-4.2 1.6t-1.7 4.1q0 1.8 0.9 3.6t2.9 4.1 3.9 4.1 5.5 5l0.2 0.2z m7.3-25.9q3.9 0 6.5 2.7t2.6 6.4q0 2.3-0.8 4.5t-3.2 4.8-4.2 4.4-6 5.6l-2.4 2.1-2.4-2.1q-5.4-4.8-7.8-7.2t-4.4-5.7-2-6.4q0-3.8 2.6-6.4t6.5-2.7q4.5 0 7.5 3.5 3-3.5 7.5-3.5z\\"></path></g></svg></div></section>"`;
